mlflow/types/schema.py

```diff
@@ -406,3 +458,280 @@ def __eq__(self, other) -> bool:
+        self,
+        name: str,
+        type: Union[DataType, str],  # pylint: disable=redefined-builtin
+        default: Union[DataType, List[DataType], None],
+        shape: Optional[Tuple[int, ...]] = None,
+    ):
+        self._name = str(name)
+        self._shape = tuple(shape) if shape is not None else None
+
+        try:
+            self._type = DataType[type] if isinstance(type, str) else type
+        except KeyError:
+            supported_types = [t.name for t in DataType if t.name != "binary"]
+            raise MlflowException(
+                f"Unsupported type '{type}', expected instance of DataType or "
+                f"one of {supported_types}",
+                INVALID_PARAMETER_VALUE,
+            )
+        if not isinstance(self.type, DataType):
+            raise TypeError(
+                "Expected mlflow.models.signature.Datatype or str for the 'type' "
+                f"argument, but got {self.type.__class__}"
+            )
+        if self.type == DataType.binary:
+            raise MlflowException(
+                f"Binary type is not supported for parameters, ParamSpec '{self.name}'"
+                "has type 'binary'",
+                INVALID_PARAMETER_VALUE,
+            )
+
+        # This line makes sure repr(self) works fine
+        self._default = default
+        self._default = self.validate_type_and_shape(repr(self), default, self.type, self.shape)
+
+    @classmethod
+    def validate_param_spec(
+        cls, value: Union[DataType, List[DataType], None], param_spec: "ParamSpec"
+    ):
+        return cls.validate_type_and_shape(
+            repr(param_spec), value, param_spec.type, param_spec.shape
+        )
+
+    @classmethod
+    def enforce_param_datatype(cls, name, value, t: DataType):
+        """
+        Enforce the value matches the data type.
+
+        The following type conversions are allowed:
+
+        1. int -> long, float, double
+        2. long -> float, double
+        3. float -> double
+        4. any -> datetime (try conversion)
+
+        Any other type mismatch will raise error.
+
+        :param name: parameter name
+        :param value: parameter value
+        :param t: expected data type
+        """
+        if value is None:
+            return value
+
+        if not np.isscalar(value):
+            raise MlflowException(
+                f"Value should be a scalar for param {name}, got {value}", INVALID_PARAMETER_VALUE
+            )
+
+        if getattr(DataType, f"is_{t.name}")(value):
+            return value
+
+        if (
+            (DataType.is_integer(value) and t in (DataType.long, DataType.float, DataType.double))
+            or (DataType.is_long(value) and t in (DataType.float, DataType.double))
+            or (DataType.is_float(value) and t == DataType.double)
+            or t == DataType.datetime
+        ):
+            try:
+                if t == DataType.datetime:
+                    return np.datetime64(value)
```

Comment:
Does this work for all precisions? Just making sure that there isn't some lossy inference operation that happens with the changes that were made to numpy last year with datetime value processing

Reply:
I think np.datetime64 is compatible with different precisions
<img width="458" alt="image" src="https://github.com/mlflow/mlflow/assets/82044803/f3fe686f-1e89-4946-9e9a-2da55bf00795">

Reply:
SGTM! Thank you for the validation :D